server/src/main/java/org/elasticsearch/action/bulk/IncrementalBulkService.java

```diff
@@ -0,0 +1,156 @@
+            if (topLevelFailure != null) {
+                assert releasables.isEmpty();
+                assert bulkRequest == null;
+                Releasables.close(releasable);
+                nextItems.run();
+            } else {
+                internalAddItems(items, releasable);
+
+                if (shouldBackOff()) {
+                    client.bulk(bulkRequest, ActionListener.runAfter(new ActionListener<>() {
+                        @Override
+                        public void onResponse(BulkResponse bulkResponse) {
+                            responses.add(bulkResponse);
+                            releaseCurrentReferences();
+                            createNewBulkRequest();
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            topLevelFailure = e;
+                            releaseCurrentReferences();
+                            bulkRequest = null;
+                        }
+                    }, nextItems));
+                } else {
+                    nextItems.run();
+                }
+            }
+        }
+
+        private boolean shouldBackOff() {
+            // TODO: Implement Real Memory Logic
+            return bulkRequest.requests().size() > 16;
+        }
+
+        public void lastItems(List<DocWriteRequest<?>> items, Releasable releasable, ActionListener<BulkResponse> listener) {
+            if (topLevelFailure != null) {
+                assert releasables.isEmpty();
+                assert bulkRequest == null;
+                releasable.close();
+                listener.onFailure(topLevelFailure);
+            } else {
+                internalAddItems(items, releasable);
+
+                client.bulk(bulkRequest, listener.delegateFailureAndWrap((l, bulkResponse) -> {
+                    responses.add(bulkResponse);
+                    releaseCurrentReferences();
+                    l.onResponse(combineResponses());
+                }));
+            }
+        }
+
+        private void internalAddItems(List<DocWriteRequest<?>> items, Releasable releasable) {
+            bulkRequest.add(items);
+            releasables.add(releasable);
+        }
+
+        private void createNewBulkRequest() {
+            bulkRequest = new BulkRequest();
+
+            if (waitForActiveShards != null) {
+                bulkRequest.waitForActiveShards(waitForActiveShards);
+            }
+            if (timeout != null) {
+                bulkRequest.timeout(timeout);
+            }
+            if (refresh != null) {
+                bulkRequest.setRefreshPolicy(refresh);
+            }
+        }
+
+        private void releaseCurrentReferences() {
+            releasables.forEach(Releasable::close);
+            releasables.clear();
+        }
+
+        private BulkResponse combineResponses() {
+            long tookInMillis = 0;
```

Comment:
The combineResponses() method iterates over the responses multiple times. This could be optimized to a single pass if performance becomes an issue with large numbers of responses.